hashintel/hash
# Make Python worker aware of the type system

## 🌟 What is the purpose of this PR?

We need the type system information in the temporal worker and preferable in a strongly typed fashion so we (a) can rely on Pydantic to parse it correctly and (b) will immediately get notified about incompatibilities.

## 🔗 Related links

- Part of [this Asana task](https://app.asana.com/0/1204727813960182/1204904869169432/f) _(internal)_

## 🚫 Blocked by

- https://github.com/hashintel/hash/pull/2693
- https://github.com/hashintel/hash/pull/2695
- https://github.com/hashintel/hash/pull/2696 (CI will fail until merged)
- https://github.com/hashintel/hash/pull/2702 (CI is likely to fail until merged)

## 🔍 What does this change?

- Add `pydantic` and configuration to the temporal worker
- Implement the type system in Python so the schemas are strongly typed
- Add simple workflow to test the conversions

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?

This PR:


### 📜 Does this require a change to the docs?

The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:


## 🛡 What tests cover this?

This is currently not tested automatically


## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  Start external services (`temporal-setup`, `hash-temporal-worker-ts`, `hash-temporal-worker-py`, `graph`)
1.  Create an account or use an existing one (here: `00000000-0000-0000-0000-000000000000`)
1.  Confirm the following workflow returns the (recursive) person entity type:

    ```
    temporal workflow execute -t aipy --type getEntityType -i '{ "entityTypeId": "https://blockprotocol.org/@examples/types/entity-type/person/v/1", "actorId": "00000000-0000-0000-0000-000000000000" }'
    ```

1. Test other URLs and ontology types

## 📹 Demo

```
Running execution:
  WorkflowId  086ec237-66a6-40be-9096-75d57848a167                                                                                                    
  RunId       af1a7f0b-41c8-4379-bc17-72690a0860d1                                                                                                    
  Type        getEntityType                                                                                                                           
  Namespace   default                                                                                                                                 
  TaskQueue   aipy                                                                                                                                    
  Args        [{"actorId":"00000000-0000-0000-0000-000000000000","entityTypeId":"https://blockprotocol.org/@examples/types/entity-type/person/v/1"}]  

Progress:
  ID          Time                          Type                  
   1  2023-06-30T12:13:08Z  WorkflowExecutionStarted              
   2  2023-06-30T12:13:08Z  WorkflowTaskScheduled                 
   3  2023-06-30T12:13:08Z  WorkflowTaskStarted                   
   4  2023-06-30T12:13:08Z  WorkflowTaskCompleted                 
   5  2023-06-30T12:13:08Z  StartChildWorkflowExecutionInitiated  
   6  2023-06-30T12:13:08Z  ChildWorkflowExecutionStarted         
   7  2023-06-30T12:13:08Z  WorkflowTaskScheduled                 
   8  2023-06-30T12:13:08Z  WorkflowTaskStarted                   
   9  2023-06-30T12:13:08Z  WorkflowTaskCompleted                 
  10  2023-06-30T12:13:08Z  ChildWorkflowExecutionCompleted       
  11  2023-06-30T12:13:08Z  WorkflowTaskScheduled                 
  12  2023-06-30T12:13:08Z  WorkflowTaskStarted                   
  13  2023-06-30T12:13:08Z  WorkflowTaskCompleted                 
  14  2023-06-30T12:13:08Z  WorkflowExecutionCompleted            

Result:
  Run Time: 1 seconds
  Status: COMPLETED
  Output: [{"$id":"https://blockprotocol.org/@examples/types/entity-type/person/v/1","$schema":"https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type","description":"An extremely simplified representation of a person or human being.","kind":"entityType","links":{"https://blockprotocol.org/@examples/types/entity-type/employed-by/v/1":{"items":{"oneOf":[{"$ref":"https://blockprotocol.org/@examples/types/entity-type/company/v/1"}]},"minItems":0,"type":"array"}},"properties":{"https://blockprotocol.org/@blockprotocol/types/property-type/name/":{"$ref":"https://blockprotocol.org/@blockprotocol/types/property-type/name/v/1"},"https://blockprotocol.org/@examples/types/property-type/e-mail/":{"$ref":"https://blockprotocol.org/@examples/types/property-type/e-mail/v/1"}},"required":["https://blockprotocol.org/@blockprotocol/types/property-type/name/"],"title":"Person","type":"object"}]
```
